src/model_theory/rings/language.lean

```diff
@@ -0,0 +1,75 @@
+/-
+Copyright (c) 2022 Joseph Hua. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Hua
+-/
+import model_theory.terms_and_formulas
+
+/-!
+# The language of rings
+
+## Main Definitions
+* A `first_order.language.ring.L` defines the language of rings,
+  which consists of `0`, `1`, `-`, `+`, `*`.
+-/
+universes u
+
+namespace first_order
+namespace language
+namespace ring
+
+/-- The constant symbols in `first_order.language.ring.L` -/
+inductive consts : Type
+| zero : consts
+| one : consts
+
+/-- The unary function symbols in `first_order.language.ring.L` -/
+inductive unaries : Type
+| neg : unaries
+
+/-- The binary function symbols in `first_order.language.ring.L` -/
+inductive binaries : Type
+| add : binaries
+| mul : binaries
+
+/-- All function symbols in `first_order.language.ring.L` -/
+def functions : ℕ → Type
+| 0 := consts
+| 1 := unaries
+| 2 := binaries
+| (n + 3) := pempty
+
+instance : inhabited consts := ⟨ consts.zero ⟩
+instance : inhabited unaries := ⟨ unaries.neg ⟩
+instance : inhabited binaries := ⟨ binaries.add ⟩
+instance : inhabited (functions 0) := ⟨ consts.zero ⟩
+
+/-- The language of rings -/
+@[reducible] def L : language :=
+{ functions := functions,
+  relations := λ n, empty }
+
+variable {α : Type u}
+
+@[simp] instance : has_zero (L.term α) := ⟨ @func L _ 0 consts.zero fin_zero_elim ⟩
```

Comment:
There is now a coercion from constants to terms - perhaps you can use it here.

Reply:
I can't figure out the right way of doing this. I am currently going with 

```
@[simp] def zero : L.constants := consts.zero

@[simp] instance : has_zero (L.term α) := ⟨ zero ⟩ 
```

Reply:
Ah cool. I'll wait for that PR then.

Reply:
It has now merged!

Reply:
this should now be working, only concern is I may have merged things incorrectly